feat: add the remainder of valid fields to Launchd

dsully opened the pull request:
More complex code for Sockets, KeepAlive, MachServices & ResourceLimits.

Let me know if you're ok with this approach and I can add some tests.

Here's a driver program I've been using to test:

```rust
use launchd::mach_services::{MachServiceEntry, MachServiceOptions};
use launchd::sockets::SocketFamily;
use launchd::{
    BonjourTypes, KeepAliveOptions, KeepAliveType, Launchd, ProcessType, ResourceLimits, Socket,
    SocketOptions, Sockets,
};
use std::collections::HashMap;
use std::path::Path;

fn main() {
    let keep_alive = KeepAliveOptions::new()
        .with_successful_exit(true)
        .with_network_state(true);

    let socket_options = SocketOptions::new()
        .with_node_name("127.0.0.1")
        .with_family(SocketFamily::Unix)
        .with_bonjour(BonjourTypes::from(vec![
            String::from("foo"),
            String::from("bar"),
        ]));

    let socket = Socket::new("Listener", socket_options);

    let mut mach_services = HashMap::new();

    mach_services.insert("com.apple.test".to_string(), MachServiceEntry::from(true));
    mach_services.insert(
        "org.rust-lang".to_string(),
        MachServiceEntry::from(MachServiceOptions::new().reset_at_close()),
    );

    let launchd = Launchd::new("LABEL", Path::new("./foo"))
        .unwrap()
        .with_user_name("Henk")
        .with_program_arguments(vec!["Hello".to_string(), "World!".to_string()])
        .disabled()
        .with_keep_alive(KeepAliveType::from(keep_alive))
        .with_process_type(ProcessType::Background)
        .with_socket(Sockets::from(socket))
        .with_soft_resource_limits(
            ResourceLimits::new()
                .with_core(1)
                .with_data(2)
                .with_file_size(3)
                .with_stack(5)
                .with_cpu(6),
        )
        .with_mach_services(mach_services)
        .with_inetd_compatibility(true);

    launchd.to_file_xml(Path::new("./test.plist")).unwrap();
}
```

dsully committed:
feat: add the remainder of valid fields to Launchd

More complex code for Sockets, KeepAlive, MachServices & ResourceLimits.

dsully committed:
Fixes:

* Move ProcessType into it's own module to handle a serde bug.

* Use Defaults.

* Backwards incompatible: .program is now Optional as either Program or ProgramArguments can exist.

koenichiwa commented:
Hi dsully, thank you for your contribution! Your approach looks great to me! You've added a lot of variables in these commits and I haven't touched this library in a long time, so I'll need some time to review every change you did. Anyways, thanks a lot, and let me know when you feel like you're ready.

dsully commented:
It's ready to review now. I'll add tests as a follow up PR or commit.

koenichiwa commented on src/keep_alive.rs:
```rust
impl KeepAliveOptions {
    pub fn new() -> Self {
        Self {
            ..Default::default()
        }
    }
...
```
Could be 
``` rust
pub fn new() -> Self {
        Default::default() // or Self::default
}
```

Leaving me to question why the `new` function exists. I mean it's fine, but could you explain your reasoning?

koenichiwa replied:
I'm looking through rust [api guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html?highlight=Default#types-eagerly-implement-common-traits-c-common-traits) and I see that it is the convention to implement an empty new when you implement Default. 

> Note that it is common and expected for types to implement both Default and an empty new constructor. new is the constructor convention in Rust, and users expect it to exist, so if it is reasonable for the basic constructor to take no arguments, then it should, even if it is functionally identical to default.

But the implementation could be simpler

dsully replied:
That + that when I was adding fields to `struct Launchd` I had to fix the direct `let launchd = Launchd { ... }` calls.

koenichiwa commented on src/mach_services.rs:
The `Launchd.MachServices` specification reads to me as: its either a dictionary of booleans, or a dictionary of dictionaries. I might be wrong, but it looks to me that with this implementation it's a dictionary of a mix of dictionaries and booleans. Subtle difference.

Also same as keep_alive.rs (the `fn new`)

dsully replied:
You're right - I found an example usage here:

https://github.com/timepigeon/clean-osx/blob/41989fc893175623a4ae5ee542f97281f38c185a/LaunchAgents/com.apple.AssistiveControl.plist#L7

Will fix.

koenichiwa replied:
Oh wow in your example it *is* a dictionary of a mix of dictionaries and booleans. So I guess your implementation was right 👍

dsully replied:
It isn't quite right, as it can't read that file currently. 😅

koenichiwa commented on src/resource_limits.rs:
Same `fn new` comment, otherwise lgtm

koenichiwa commented on src/sockets.rs:
I'm not sure if SockFamily::Unix is a good idea. I get where you got it from:

> SockFamily <string>
           This optional key can be used to specifically request that "IPv4"
           or "IPv6" socket(s) be created.

> SockPathName <string>
           This optional key implies SockFamily is set to "Unix". It specifies
           the path to [connect(2)](https://www.manpagez.com/man/2/connect/) or [bind(2)](https://www.manpagez.com/man/2/bind/) to.

But I see only that it can be implied to be Unix, not set to. It could be that this value could be any of the [socket address families](https://man7.org/linux/man-pages/man7/address_families.7.html) though, in which case the enum is too restrictive.

koenichiwa replied:
Shouldn't BonjourTypes be BonjourType?

dsully replied:
> 

Agreed, and I was also hesitant to do it that way. It's unclear of any of those socket families can be used. By the way the man page (`launchd.plist(5)`) is worded, it doesn't seem like it. I can remove if you prefer.

koenichiwa replied:
Maybe we can device a test to check which inputs are valid. Or, at least check if "Unix" is valid, and then we can create a new issue to extend the enum.

koenichiwa commented on src/lib.rs:
`Launchd.program` is changed to `Option<String>`, which falls in the specs of launchd. But, if `program_arguments` is empty or None, then the struct is invalid. You could prevent the creation of an invalid struct by demanding at least one in the `new` function, but that doesn't work anymore due to `derive(Default)`. Another solution might be something like a [builder pattern with a typestate](https://www.greyblake.com/blog/builder-with-typestate-in-rust/)

It's not a hard requirement for this PR yet, but maybe you have a suggestion how to keep the struct valid at all times?

koenichiwa replied:
I had a moment where I couldn't figure out what umask is, but it's basically unix permissions. Due to that moment I'm not sure if the type u16 is descriptive enough, but maybe just a comment would help.

koenichiwa replied:
Curious if the type of `inetdCompatibility` is right. `String` in `Option<HashMap<String, bool>>` might be too broad. Shouldn't it be an enum of Wait (and NoWait)??

koenichiwa replied:
The variables in the `Launchd` struct were in the same order as the [manual](https://www.manpagez.com/man/5/launchd.plist/) to check if we didn't forget any variables. Now this is true up until `on_demand`

dsully replied:
Re: `program` and `program_arguments` - I went down a number of paths on this, including having it as a `ProgramTypes` option, a custom `serde` Serializer (on the `struct Launchd`) and was never happy with any of those. I'll take a look at that pattern. Any ideas here are welcome.

Re: `inetdCompatibility` - sounds good.

Will fix the ordering.

koenichiwa replied:
> Re: program and program_arguments - I went down a number of paths on this, including having it as a ProgramTypes option, a custom serde Serializer (on the struct Launchd) and was never happy with any of those. I'll take a look at that pattern. Any ideas here are welcome.

I'll think about other solutions. But as I said, ensuring validity might be outside the scope of this PR. I'll come back to you on this.

> inetdCompatibility - sounds good.

Great, just check if NoWait is also valid, or that the only valid key is Wait. I only found examples with Wait up until now.

Also thanks for fixing the ordering.

koenichiwa replied:
One other solution might be to create a validation function? From what I remember, the only requirements where that the `label` was populated (I'm not sure if Mac is going to be happy with empty strings) and that either the `program` holds the path or the `program_args`

```rust
// NB: two minute draft
fn validate(&self) -> Result<(), Error> {
    if self.label.is_empty() {
        return Err(Error::InvalidState);
    }
    if self.program.is_none() {
        if let Some(args) = &self.program_arguments {
            if args.is_empty() {
                return Err(Error::InvalidState);
            }
        } else {
            return Err(Error::InvalidState);
        }
    }
    
    Ok(())
}
```

The problem is finding out where this function should be called... I think this is out of scope, keep it how it is and I'll create a new issue when this feature is merged.

dsully replied:
Yeah - I think on Serialization is the only place.

koenichiwa replied:
Looks good! I saw you added some extra variables. Also some that are deprecated, for reading purposes, which is good! I missed the ServiceIPC variable. I get that the description tells you not to use it:

>[ServiceIPC <boolean>](https://keith.github.io/xcode-man-pages/launchd.plist.5.html#ServiceIPC)
    Please remove this key from your launchd.plist.

 But maybe we should add it for reading purposes as well.

Oh and I was curious, I've seen the trait Eq being added and removed a couple of times if I remember correctly. What was the reason?

Otherwise, great!!

dsully replied:
Once I added `plist::Value` in the `type` for `LaunchEvents`, it doesn't implement `Eq`, so I had to remove it again.

koenichiwa reviewed (commented):
Hey this looks really great! Of course have some nitpicks/questions. The one I'm the most curious about is about MachServices.

Also, I'm looking forward to those tests :)

dsully committed:
Remaining Review Comments + Additional Fields

As I was checking the order of fields in launchd.plist(5), I noticed some missing fields and the incorrect definition of LoadToSessionType.

koenichiwa reviewed (commented):
Thanks for the changes!

dsully committed:
Add ServiceIPC and the some tests + test data files.

dsully commented:
Let me know when you have time for another pass at this. Thanks

dsully commented:
Gentle nudge. 😄

dsully commented:
Ping.. on summer holiday?

koenichiwa commented:
Oh my god I'm so sorry, no I was busy with my thesis, and it must've slipped my mind. I'll do this first thing today.

koenichiwa commented:
Hey I did a pass over this. It looks good to me. I'll merge and put out a new version. I'm really sorry for the wait

dsully commented:
Thanks - no problem, figured you were busy with life.

I'll take a look at the `program` / `program_arguments` when I have some time.

koenichiwa commented:
> I'll take a look at the `program` / `program_arguments` when I have some time.

Alright thanks! I'm also looking into options of deriving a builder type. I haven't found a suitable choice yet. Currently the best option is [typed-builder](https://crates.io/crates/typed-builder), but it doesn't seem very mature and still comes with a lot of issues. I'll create a specific issue for this problem.

I also needed to create a "bandaid" for launchevents. It wouldn't work with every feature combination. I should've found this before merging, but alas, I thought the amount of work you've done warranted a new version for now.
